src/NuGet.Services.AzureSearch/SearchService/SearchTextBuilder.cs

```diff
@@ -97,36 +94,50 @@ public string Build(ParsedQuery parsed)
                 return MatchAllDocumentsQuery;
             }
 
-            // Generate a Lucene query for Azure Search.
-            var builder = new AzureSearchQueryBuilder();
             var scopedTerms = parsed.Grouping.Where(g => g.Key != QueryField.Any && g.Key != QueryField.Invalid).ToList();
             var unscopedTerms = parsed.Grouping.Where(g => g.Key == QueryField.Any)
                 .Select(g => g.Value)
-                .SingleOrDefault();
+                .SingleOrDefault()?
+                .ToList();
 
-            var requireScopedTerms = unscopedTerms != null
-                || scopedTerms.Select(t => t.Key).Distinct().Count() > 1;
+            // Don't bother generating Azure Search text if all terms are scoped to invalid fields.
+            var hasUnscopedTerms = unscopedTerms != null && unscopedTerms.Count > 0;
+            if (scopedTerms.Count == 0 && !hasUnscopedTerms)
+            {
+                return MatchAllDocumentsQuery;
+            }
+
+            // Add the terms that are scoped to specific fields.
+            var builder = new AzureSearchTextBuilder();
+            var requireScopedTerms = hasUnscopedTerms || scopedTerms.Count > 1;
 
             foreach (var scopedTerm in scopedTerms)
             {
                 var fieldName = FieldNames[scopedTerm.Key];
                 var values = ProcessFieldValues(scopedTerm.Key, scopedTerm.Value).ToList();
 
-                builder.AppendScopedTerms(fieldName, values, required: requireScopedTerms);
+                if (values.Count > 1)
+                {
+                    builder.AppendScopedTerms(fieldName, values, required: requireScopedTerms);
+                }
+                else
+                {
+                    builder.AppendScopedTerm(fieldName, values.First(), required: requireScopedTerms);
+                }
             }
 
-            if (unscopedTerms != null)
+            // Add the terms that can match any fields.
+            if (hasUnscopedTerms)
             {
-                builder.AppendTerms(unscopedTerms.ToList());
-            }
+                builder.AppendTerms(unscopedTerms);
 
-            var result = builder.ToString();
-            if (string.IsNullOrWhiteSpace(result))
-            {
-                return MatchAllDocumentsQuery;
+                if (unscopedTerms.Count > 1)
+                {
+                    builder.AppendBoostIfMatchAllTerms(unscopedTerms, _options.Value.MatchAllTermsBoost);
```

Comment:
Could you leave a comment explaining why we don't boost the scoped terms too?

Reply:
Done